[TG-3694] Parse and store inner class information

jeannielynnmoulton opened the pull request:
This parses the InnerClasses attribute of the java bytecode with the intention of determining
- If an parsed class is an inner class
- The accessibility information for inner classes

If the parsed class is an inner class, then the original accessibility is overwritten by the information from the InnerClasses attribute.

Whether the class is an inner class is then stored in java_class_typet.

...

@thk123 @thomasspriggs Currently there is one test failing in test gen with the changes that will require some investigation, but I am on AL Friday and wanted to get some preliminary reviews so with some luck, this might be finished the sprint. Thanks.

@thk123 Could you please add an in progress label or equivalent. I have no privileges.

Commit:
Parses InnerClasses attribute of java bytecode.

In this commit, nothing is done with the data.

Commit:
Stores inner class data in java class types.

Though this data is stored, it will not be used in test-gen yet because test-gen is assuming all non-public inner classes are private.

Commit:
Adds unit test for parsing inner classes.

thomasspriggs commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1640,6 +1640,48 @@ void java_bytecode_parsert::rclass_attribute(classt &parsed_class)
     read_bootstrapmethods_entry(parsed_class);
   }
+  else if(attribute_name == "InnerClasses")
+  {
```
This section should be moved into a new function/method, to avoid `rclass_attribute` getting too long.

thomasspriggs commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1640,6 +1640,48 @@ void java_bytecode_parsert::rclass_attribute(classt &parsed_class)
+      "The number of bytes to be read for the InnerClasses attribute does not "
+      "match the attribute length.");
+
+    const std::function<pool_entryt &(u2)> pool_entry_lambda =
```
NIT: As this `pool_entry_lambda` is only used once, I would move it into the call to `.get_name`, rather than storing it in a variable.

thk123 replied:
I should probably add this exception to the coding standard, but I think `auto` is preferable for a lambda for two reasons:
1. By putting it into a `std::function` you allow the value to be null whereas the `auto` allows the compiler to pick a precise lambda type 
2. Probably doesn't help readability as the inputs and outputs are repeated on the rhs (`[this](u2 index) -> pool_entry &`)

thomasspriggs commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1640,6 +1640,48 @@ void java_bytecode_parsert::rclass_attribute(classt &parsed_class)
+      UNUSED u2 inner_name_index = read_u2();
+      u2 inner_class_access_flags = read_u2();
+
+      if(inner_class_info_index != 0)
```
Bike shedding: I would prefer `if(inner_class_info_index == 0) continue;` over indenting the remainder of the loop. I think opinions on this one are likely to vary.

jeannielynnmoulton replied:
~I've been asked to do worse. I am not sure which is clearer, but I've changed it~ 👍  ~No I haven't because the bytes need to be read anyway.~ No I am being stupid, changed :) Can you do double strikethrough?

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1640,6 +1640,48 @@ void java_bytecode_parsert::rclass_attribute(classt &parsed_class)
+
+    const std::function<pool_entryt &(u2)> pool_entry_lambda =
+      [this](u2 index) -> pool_entryt & { return pool_entry(index); };
+    std::function<std::string(std::string, char)> remove_separator_char =
```
Nit: `const`

thomasspriggs commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1640,6 +1640,48 @@ void java_bytecode_parsert::rclass_attribute(classt &parsed_class)
+    u2 number_of_classes = read_u2();
+    u4 number_of_bytes_to_be_read = number_of_classes * 8 + 2;
+    INVARIANT(
+      number_of_bytes_to_be_read == attribute_length,
```
It is good that we have this check. However I think this code could be more robust, if we have the condition `number_of_bytes_to_be_read <= attribute_length` and a call to `skip_bytes` for the difference at the end of this block.

jeannielynnmoulton replied:
I see what you are saying, but I am not sure that it's best. If the bytes don't match, I don't think we can guarantee that anything makes sense because that would mean that the bytecode spec has changed basically, and so should probably just fail hard.

thomasspriggs commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1669,16 +1669,23 @@ void java_bytecode_parsert::rclass_attribute(classt &parsed_class)
         // the parsed class is an inner class, so overwrite the parsed class'
         // access information and mark it as an inner class
-        UNUSED bool is_inner_class =
+        bool is_inner_class =
```
You could use `parsed_class.is_inner_class = ` here instead of declaring an additional local variable and hence save a line of code inside the if block below.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1640,6 +1640,48 @@ void java_bytecode_parsert::rclass_attribute(classt &parsed_class)
+        UNUSED bool is_protected = inner_class_access_flags & ACC_PROTECTED;
+
+        // If the original parsed class name matches the inner class name
+        // the parsed class is an inner class, so overwrite the parsed class'
```
Why does this work? Also possibly I think you might be able to check whether `outer_class_info_index == 0`? 

> If C is not a member of a class or an interface (that is, if C is a top-level class or interface (JLS §7.6) or a local class (JLS §14.3) or an anonymous class (JLS §15.9.5)), the value of the outer_class_info_index item must be zero.

Though that sounds like anonymous and local classes would be marked as not inner, which I guess is true?

thomasspriggs replied:
Anonymous classes and local classes are both inner classes. If they are not being marked as such then we have a problem. I think the accessibility for these should always be private.

thomasspriggs replied:
Thinking about this, it may be advantageous to have a is_anonymous / is_local flag set at this point. We currently have an `is_anonymous` method in `state_typet`, which attempts to work this out based on regexing the name, which isn't the ideal approach.

jeannielynnmoulton replied:
The problems on test-gen were related to treating anonymous/local classes the same as inner classes. This has been updated by checking that outer_class_info_index  is 0 or not (it's 0 when it is outer, interface, local or anonymous). @thk123 The reason we can't just check outer_class_info_index is because the we only want to overwrite the accessibility information when the current parsed class (not the current parsed inner class information) is an inner class. As far as I can tell, there is no magic way to do this from the spec.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1640,6 +1640,48 @@ void java_bytecode_parsert::rclass_attribute(classt &parsed_class)
   }
+  else if(attribute_name == "InnerClasses")
+  {
+    u2 number_of_classes = read_u2();
```
You might consider pulling this out into a `read_innerclass_attribute` which would allow you to document when you'd expect to see this and an overview of the info we're getting out of it, as I can't get my head around when this attribute appears and what we find in it, despite reading the spec.

jeannielynnmoulton replied:
The spec is not great, agreed.

thk123 commented on jbmc/src/java_bytecode/java_types.h:
```diff
@@ -111,6 +111,16 @@ class java_class_typet:public class_typet
     return set(ID_access, access);
   }
 
+  const irep_idt &get_inner_class() const
```
Return a bool (using `get_bool` as done in `get_final`) as this hides the messiness that everything is strings from the user of this function.

thomasspriggs replied:
Could these methods be called `get_is_inner_class` and `set_is_inner_class` instead? The calls to the current method names superficially appear to be a call to a method on an outer class to get an inner class belonging to it.

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java:
```diff
@@ -0,0 +1,26 @@
+            this.i = i;
+        }
+    }
+}
```
Nit: missing new line

thomasspriggs replied:
FYI: The clion/intellij settings can be tweaked to automatically add a new line to the end of files when it is missing. Should save a bunch of time on comments around this in your PRs.

jeannielynnmoulton replied:
Added and configured. 👍

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp:
```diff
@@ -0,0 +1,84 @@
+      THEN("The class should be marked as public")
+      {
+        const symbolt &class_symbol =
+          *new_symbol_table.lookup("java::InnerClasses$PublicInnerClass");
```
Nit: use `lookup_ref` if you are going to immediately dereference the return from `lookup` as it will produce a nicer error if the symbol doesn't exist.

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java:
```diff
@@ -0,0 +1,26 @@
+public class InnerClasses {
```
Could you add a non-public class to this file (in case that means there are two inner class attributes?) Unless that isn't true, in which case might not be worth it

jeannielynnmoulton replied:
Added a non-public outer class 👍

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp:
```diff
@@ -0,0 +1,84 @@
+          *new_symbol_table.lookup("java::InnerClasses$PublicInnerClass");
+        const java_class_typet java_class =
+          to_java_class_type(class_symbol.type);
+        REQUIRE(java_class.get_inner_class() == ID_is_inner_class);
```
I think this is wrong, I think the string will be something like "1" or "true" - which probably means if this passes CI these unit tests aren't being run.

thk123 replied:
Indeed, it is the missing from Makefile as the CMake build fails:
https://travis-ci.org/diffblue/cbmc/jobs/397934975#L7792
`1 == is_inner_class`

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp:
```diff
@@ -0,0 +1,84 @@
+  {
+    WHEN("Parsing the InnerClasses attribute for a public inner class")
+    {
+      const symbol_tablet &new_symbol_table =
```
Since you always load the same file, I'd probably do this at the top of scenario to reduce duplication

thk123 commented on jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp:
```diff
@@ -0,0 +1,84 @@
+/*******************************************************************\
```
Missing from the `Makefile`.

jeannielynnmoulton replied:
I am not sure what this means or how to add it. I will ask you when you are done with your current conversation.

thk123 reviewed (commented):
Looks pretty good - couple of suggestions and I think your tests need tweaking.

thomasspriggs commented on jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java:
```diff
@@ -0,0 +1,26 @@
+        private PrivateInnerClass(int i) {
+            this.i = i;
+        }
+    }
```
Could you extend this test with some Inner Inner class examples, to show that things still work when the classes are nested an additional layer deep?

thomasspriggs replied:
Could you also extend this with an anonymous class and a local class?

thomasspriggs commented on jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp:
```diff
@@ -0,0 +1,84 @@
+    WHEN("Parsing the InnerClasses attribute for a public inner class")
+    {
+      const symbol_tablet &new_symbol_table =
+        load_java_class("InnerClasses", "./java_bytecode/java_bytecode_parser");
```
Can we also have some tests where an inner class is loaded first, instead of an outer class containing inner classes?

thomasspriggs reviewed (commented):
Some nit picking. Couple of requests for extensions to the unit test.

Commit:
Refactors parsing of inner classes attribute.

thomasspriggs commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1576,6 +1577,65 @@ void java_bytecode_parsert::relement_value_pair(
+        .get_name(pool_entry_lambda);
+    bool is_private = inner_class_access_flags & ACC_PRIVATE;
+    bool is_public = inner_class_access_flags & ACC_PUBLIC;
+    bool is_protected = inner_class_access_flags & ACC_PROTECTED;
```
NIT: These three lines give a clang-tidy warning in clion, due to the implicit cast from integer to bool. This could be fixed as follows - `bool is_protected = (inner_class_access_flags & ACC_PROTECTED) != 0;`

thk123 replied:
Or a `read_bitflag` method but I wouldn't block this PR on it since this is a common approach used in this file.

thomasspriggs replied:
Looking through the rest of this file, all the other instances of this have already been fixed to avoid the implicit cast.

jeannielynnmoulton replied:
Changed

thomasspriggs replied:
I considered it to be an optional "NIT" comment. I guess it would have been helpful if @thk123 had commented with something along the lines of "This is not a NIT, fix it."

Commit:
Fixes parsing for anonymous classes

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a129281e9ec21a24aeb1a228ca5096a5b0e7c867).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77913660

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: bccf9cc217d6944b919586733a445f750ebd598b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77914476

jeannielynnmoulton commented:
@thk123 @thomasspriggs This is ready for re-review. Seems to have passed the compatibility check last night, which is what I would expect now that it is treating anonymous/local classes correctly and test-gen is still ignoring most of inner classes, so I'm reasonably confident that PR2014 will pass too. I doubt the test-gen changes one will make it into the sprint, but it's here anyway for reference.

Test-gen with changes: https://github.com/diffblue/test-gen/pull/2012

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1576,6 +1578,79 @@ void java_bytecode_parsert::relement_value_pair(
+    UNUSED u2 inner_name_index = read_u2();
+    u2 inner_class_access_flags = read_u2();
+
+    if(inner_class_info_index == 0)
```
Comment for why would be good here? My reading of the spec is it _can't_ be zero?

thomasspriggs replied:
Could still do with a comment here.

jeannielynnmoulton replied:
Removed

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: c0c1029208a62433a3803b5aebc727337b22b31c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77954988

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1576,6 +1578,79 @@ void java_bytecode_parsert::relement_value_pair(
+    {
+      // This is a marker for an anonymous or local class
+      // which are treated as private
+      parsed_class.is_private = true;
```
Sorry if spotted this before, but having three booleans to represent a single value is error prone, I'd prefer an enum in the `classt`

thomasspriggs replied:
This is a common approach in this file _and_ it effectively matches the java bytecode.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1576,6 +1578,79 @@ void java_bytecode_parsert::relement_value_pair(
+/// Corresponds to the element_value structure
+/// Described in Java 8 specification 4.7.6
+/// https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.6
+/// Parses the bytes of the InnerClasses attribute for the current parsed class,
```
Nit: fewer words "Parses ~the bytes of~ the InnerClass attribute"

thk123 replied:
Clarity: current parsed class or current parsed class file? (If the former, ignore me)

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1576,6 +1578,79 @@ void java_bytecode_parsert::relement_value_pair(
+/// Described in Java 8 specification 4.7.6
+/// https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.6
+/// Parses the bytes of the InnerClasses attribute for the current parsed class,
+/// which contains any array of information about inner classes. We are
```
Clarity: Inner classes of that class?

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1576,6 +1578,79 @@ void java_bytecode_parsert::relement_value_pair(
+/// in its inner class array. If the parsed class is not an inner class, then it
+/// is ignored. When a parsed class is an inner class, the accessibility
+/// information for the parsed class is overwritten, and the parsed class is
+/// marked as an inner class. Note that is outer_class_info_index == 0, the
```
Probably stick this note by the code it relates to rather than here.

thk123 reviewed (approved):
All comments look to be addressed, couple of suggestions for making the comment on `rinner_classes_attribute` even better. I also think the enum for accessibility of `classt` would be a big win.

Commit:
Clarifies language in documentation.

Commit:
Marks anonymous classes as inner classes

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 6ce7b1331c764d34c18f3402cfb27e0d58947c88).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77989278

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 9ba55e20aad17e94ac1fa6873681c20b3e3194c1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77998632

thk123 commented:
On a subsequent PR you said you're just waiting on CI, but I think these two comments are outstanding: 
https://github.com/diffblue/cbmc/pull/2493/files#r199757047 (documenting the inner_class_index == 0 if)
https://github.com/diffblue/cbmc/pull/2493#discussion_r199468463 (using `!= 0` rather than implicit conversion to `bool`

Commit:
Refactor inner class parsing.

jeannielynnmoulton commented:
@thk123 I have removed the `if(inner_class_index == 0)` line, so that is no longer applicable. When someone says `NIT:` I (maybe wrongly) take that as optional, but I have changed it since I am pushing anyway.

thk123 commented:
Agree NIT means optional, but since there was another change (the docs, or as it turns out, removing the if) thought I'd draw your attention to it :slightly_smiling_face:. Will merge when passes CI.

jeannielynnmoulton commented:
Waiting for CI to pass on bump: https://github.com/diffblue/test-gen/pull/2012

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 5350133c8665f7f3f65190f6f51807e6296b3198).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78105793
